packages/fuels-programs/src/script_calls.rs

```diff
@@ -242,11 +264,28 @@ where
             .map_err(|err| map_revert_error(err, &self.log_decoder))
     }
 
-    pub async fn submit(mut self) -> Result<ScriptCallHandler<T, D>> {
+    pub async fn submit(mut self) -> Result<SubmitResponse<T, D>> {
         let tx = self.build_tx().await?;
-        self.cached_tx_id = Some(self.provider.send_transaction(tx).await?);
+        let provider = self.account.try_provider()?;
+
+        let should_retry_fn =
+            |res: &Result<TxId>| -> bool { matches!(res, Err(Error::IOError(_))) };
+
+        self.cached_tx_id = Some(if self.retry_config.max_attempts != 0 {
+            retry(
+                || async { provider.send_transaction(tx.clone()).await },
+                &self.retry_config,
+                should_retry_fn,
+            )
+            .await?
+        } else {
+            provider.send_transaction(tx).await?
+        });
```

Comment:
take it if you like it:
```suggestion
        self.cached_tx_id = match self.retry_config.max_attempts {
            0 => provider.send_transaction(tx).await,
            _ => {
                retry(
                    || async { provider.send_transaction(tx.clone()).await },
                    &self.retry_config,
                    should_retry_fn,
                )
                .await
            }
        }
        .map(Some)?;
```